Asgard/Communications/Classes/GridConnectProcessor.cs

```diff
@@ -0,0 +1,166 @@
+﻿using Microsoft.Extensions.Logging;
+using System;
+using System.Buffers;
+using System.IO.Pipelines;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Asgard.Communications
+{
+    public class GridConnectProcessor:IGridConnectProcessor,IDisposable
+    {
+        private readonly ILogger<GridConnectProcessor> logger;
+
+        public ITransport Transport { get; }
+
+        public event EventHandler<TransportErrorEventArgs> TransportError;
+        public event EventHandler<MessageReceivedEventArgs> GridConnectMessage;
+        private CancellationTokenSource cts;
+        private bool disposedValue;
+
+        public GridConnectProcessor(ITransport transport, ILogger<GridConnectProcessor> logger = null)
+        {
+            Transport = transport;
+            this.logger = logger;
+        }
+
+        public void Open()
+        {
+            Transport.Open();
+            cts = new CancellationTokenSource();
+            var pipe = new Pipe();
+            //TODO: would this be better as distinct threads rather than using up threads from the threadpool?
+            ReadPipe(pipe.Reader).AsgardFireAndForget();
+            Listen(pipe.Writer).AsgardFireAndForget();
+        }
+
+        private Task Listen(PipeWriter writer)
+        {
+            return Task.Run(async () => {
+                const int minBufferSize = 64;
+                while (!cts.IsCancellationRequested)
+                {
+                    var memory = writer.GetMemory(minBufferSize);
+                    try
+                    {
+                        var read = await Transport.ReadAsync(memory, cts.Token);
+                        if (read == 0)
+                        {
+                            //todo?
+                        }
+
+                        writer.Advance(read);
+                        await writer.FlushAsync(cts.Token);
+                    }
+                    catch (TaskCanceledException)
+                    {
+                        //ok
+                    }
+                    catch (Exception e)
+                    {
+                        TransportError?.Invoke(this, new TransportErrorEventArgs(new TransportException("Error receiving bytes", e)));
+                    }
+                }
+            });
+        }
+
+        private Task ReadPipe(PipeReader reader)
+        {
+            return Task.Run(async () => {
+                while (!cts.IsCancellationRequested)
+                {
```

Comment:
Should really be checking the CancellationToken, not the CancellationTokenSource.